### PR TITLE
feat: Add multilabel classification for training

### PR DESCRIPTION
**Pull request by Pringled**
This PR adds multilabel classification to the training module. If labels are passed as a `list[list[str]]`, the model is automatically trained as a multilabel classification model.

### CODE REVIEW

File: model2vec/train/README.md [hunk 1]
```diff
@@ -65,6 +67,54 @@ print(f"Took {int((perf_counter() - s) * 1000)} milliseconds for {len(test)} ins
 # Took 67 milliseconds for 2000 instances on CPU.
 ```
 
+## Multi-label classification
+
+Multi-label classification is supported out of the box. Just pass a list of multi-labels to the `fit` function, and a multi-label classifier will be trained. For example, the following code trains a multi-label classifier on the [go_emotions](https://huggingface.co/datasets/google-research-datasets/go_emotions) dataset:
```
Comment (stephantul) [inline model2vec/train/README.md:72]: a list of multi-labels is a bit unclear.

File: model2vec/train/README.md [hunk 2]
```diff
@@ -65,6 +67,54 @@ print(f"Took {int((perf_counter() - s) * 1000)} milliseconds for {len(test)} ins
...
+print(f"Precision: {metrics.precision_score(y_true, y_pred, average='macro', zero_division=0):.3f}")
+print(f"Recall: {metrics.recall_score(y_true, y_pred, average='macro', zero_division=0):.3f}")
+print(f"F1: {metrics.f1_score(y_true, y_pred, average='macro', zero_division=0):.3f}")
+# Accuracy: 0.410
+# Precision: 0.527
+# Recall: 0.410
+# F1: 0.439
+```
+
+The scores are competetive with the popular [roberta-base-go_emotions](https://huggingface.co/SamLowe/roberta-base-go_emotions) model, while our model is orders of magnitude faster.
```
Comment (stephantul) [inline model2vec/train/README.md:116]:
```suggestion
The scores are competitive with the popular [roberta-base-go_emotions](https://huggingface.co/SamLowe/roberta-base-go_emotions) model, while our model is orders of magnitude faster.
```

File: model2vec/train/classifier.py [hunk 3]
```diff
@@ -66,14 +69,32 @@ def construct_head(self) -> nn.Sequential:
...
+        :return: The predictions.
+        """
+        pred = []
         for batch in trange(0, len(X), batch_size, disable=not show_progress_bar):
             logits = self._predict_single_batch(X[batch : batch + batch_size])
-            pred.extend([self.classes[idx] for idx in logits.argmax(1)])
-
-        return np.asarray(pred)
+            if self.multilabel:
+                probs = torch.sigmoid(logits)
```
Comment (stephantul) [inline model2vec/train/classifier.py:91]: I think this could be clearer by first turning everything into a booltensor

```
above_threshold = probs > threshold
```

```
for sample in above_threshold:
    sample_labels = [self.classes[x] for x in torch.nonzero(sample).squeeze()]
```

If we turn classes into an array, we could also index immediately:

```
self.classes[torch.nonzero(sample).squeeze()]
```

Anyway, just a small optimization

File: model2vec/train/classifier.py [hunk 4]
```diff
@@ -127,22 +155,24 @@ def fit(
         """
         pl.seed_everything(_RANDOM_SEED)
         logger.info("Re-initializing model.")
-        self._initialize(y)
+
+        # Determine whether the task is multilabel based on the type of y.
+        multilabel = isinstance(y[0], list)
```
Comment (stephantul) [inline model2vec/train/classifier.py:160]: Maybe do some additional checks:
* I don't mind checking all input items
* Users might pass lists of tuples, which will crash, but which might work? (it violates our typing, but could work, not sure what you think should happen)

File: model2vec/train/classifier.py [hunk 5]
```diff
@@ -127,22 +155,24 @@ def fit(
...
+        self._initialize(y, multilabel=multilabel)
 
         train_texts, validation_texts, train_labels, validation_labels = self._train_test_split(
-            X, y, test_size=test_size
+            X, y, test_size=test_size, multilabel=multilabel
         )
 
         if batch_size is None:
-            # Set to a multiple of 32
```
Comment (stephantul) [inline model2vec/train/classifier.py:137]: why remove this comment?

File: model2vec/train/classifier.py [hunk 6]
```diff
@@ -186,45 +216,82 @@ def fit(
...
+        Sets the output dimensionality, the classes, and initializes the head.
 
+        :param y: The labels.
+        :param multilabel: Whether the task is multilabel.
+        """
+        self.multilabel = multilabel
+        if multilabel:
+            classes = sorted({label for sublist in y for label in sublist})
+        else:
+            classes = sorted(set(cast(list[str], y)))
```
Comment (stephantul) [inline model2vec/train/classifier.py:232]: awoeee

File: model2vec/train/classifier.py [hunk 7]
```diff
@@ -186,45 +216,82 @@ def fit(
...
+    def _initialize(self, y: list[str] | list[list[str]], multilabel: bool = False) -> None:
+        """
+        Sets the output dimensionality, the classes, and initializes the head.
 
+        :param y: The labels.
+        :param multilabel: Whether the task is multilabel.
+        """
+        self.multilabel = multilabel
+        if multilabel:
+            classes = sorted({label for sublist in y for label in sublist})
```
Comment (stephantul) [inline model2vec/train/classifier.py:230]: easier:

```
from itertools import chain

classes = sorted(set(chain.from_iterable(y)))
```

File: model2vec/train/classifier.py (hunk 6 above)
Comment (stephantul) [reply model2vec/train/classifier.py:232]: I'd solve this by removing the boolean argument. This is a code smell. You let your choice of multilabel depend on a flag, but all the information to decide whether you are dealing with multilabel is here: you can simply see if you have a list of lists or not. That also solves the typing issue.

File: model2vec/train/classifier.py [hunk 8]
```diff
@@ -186,45 +216,82 @@ def fit(
...
         tokenized: list[list[int]] = [
             encoding.ids[:max_length] for encoding in self.tokenizer.encode_batch_fast(X, add_special_tokens=False)
         ]
-        labels_tensor = torch.Tensor([self.classes.index(label) for label in y]).long()
-
+        if multilabel:
+            num_classes = len(self.classes)
+            label_list = []
+            for sample_labels in y:
+                multi_hot = torch.zeros(num_classes, dtype=torch.float)
```
Comment (stephantul) [inline model2vec/train/classifier.py:263]: isn't this float by default?

File: model2vec/train/classifier.py [hunk 9]
```diff
@@ -186,45 +216,82 @@ def fit(
...
+                for label in sample_labels:
+                    index = self.classes.index(label)
```
Comment (stephantul) [inline model2vec/train/classifier.py:265]: I'd wager that calling index a whole bunch of times is pretty slow. Better make a map before. Similarly, assigning to an array multiple times is slower than doing it once.

```
# On top of function.
mapping = {label: idx for idx, label in self.classes}

# Here
multi_hot[[mapping[label] for label in sample_labels]] = 1.0
```

File: model2vec/train/classifier.py (hunk 8 above)
Comment (stephantul) [reply model2vec/train/classifier.py:263]: I also think it might be faster to just construct the whole tensor at once. So `multi_hot` = `torch.zeros(len(y), num_classes)`, and then you simply assign rows.

File: model2vec/train/classifier.py [hunk 10]
```diff
@@ -186,45 +216,82 @@ def fit(
...
-        X: list[str], y: list[str], test_size: float
-    ) -> tuple[list[str], list[str], list[str], list[str]]:
-        """Split the data."""
-        label_counts = Counter(y)
-        if min(label_counts.values()) < 2:
-            logger.info("Some classes have less than 2 samples. Stratification is disabled.")
+        X: list[str],
+        y: list[str] | list[list[str]],
+        test_size: float,
+        multilabel: bool = False,
```
Comment (stephantul) [inline model2vec/train/classifier.py:278]: You can derive multilabel from the labels, so no need to pass it.

File: model2vec/train/classifier.py [hunk 11]
```diff
@@ -186,45 +216,82 @@ def fit(
...
     @staticmethod
     def _train_test_split(
-        X: list[str], y: list[str], test_size: float
-    ) -> tuple[list[str], list[str], list[str], list[str]]:
-        """Split the data."""
-        label_counts = Counter(y)
-        if min(label_counts.values()) < 2:
-            logger.info("Some classes have less than 2 samples. Stratification is disabled.")
+        X: list[str],
+        y: list[str] | list[list[str]],
```
Comment (stephantul) [inline model2vec/train/classifier.py:276]: This should be a TypeVar, also used in the output type signature.

e..g
```
LabelType = TypeVar("LabelType", list[str], list[list[str]])

y: LabelType
) -> tuple[list[str], list[str], LabelType, LabelType]
```

File: model2vec/train/classifier.py [hunk 12]
```diff
@@ -186,45 +216,82 @@ def fit(
...
+
+        For single-label classification, stratification is attempted (if possible).
+        For multilabel classification, a random split is performed.
+        """
+        if not multilabel:
+            label_counts = Counter(y)
+            if min(label_counts.values()) < 2:
+                logger.info("Some classes have less than 2 samples. Stratification is disabled.")
+                return train_test_split(X, y, test_size=test_size, random_state=42, shuffle=True)
+            return train_test_split(X, y, test_size=test_size, random_state=42, shuffle=True, stratify=y)
```
Comment (stephantul) [inline model2vec/train/classifier.py:291]: We can approximate stratification somewhere in the future.

File: model2vec/train/classifier.py [hunk 13]
```diff
@@ -266,28 +333,35 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.model(x)
 
     def training_step(self, batch: tuple[torch.Tensor, torch.Tensor], batch_idx: int) -> torch.Tensor:
-        """Simple training step using cross entropy loss."""
+        """Training step using cross-entropy loss for single-label and binary cross-entropy for multilabel training."""
         x, y = batch
         head_out, _ = self.model(x)
-        loss = nn.functional.cross_entropy(head_out, y).mean()
-
+        if self.model.multilabel:
```
Comment (stephantul) [inline model2vec/train/classifier.py:339]: I'd rather we define this in the model, because checking this if makes it look weird.

**Review by stephantul — CHANGES_REQUESTED** [review]

File: model2vec/train/classifier.py (hunk 5 above)
Comment (Pringled) [reply model2vec/train/classifier.py:137]: My bad, I was copypasting stuff over and this got lost Iguess, added it back

File: model2vec/train/classifier.py (hunk 4 above)
Comment (Pringled) [reply model2vec/train/classifier.py:160]: Will allow tuples, good suggestion

File: model2vec/train/classifier.py (hunk 13 above)
Comment (Pringled) [reply model2vec/train/classifier.py:339]: What do you mean, define it in the model? This is in the model no?

File: model2vec/train/classifier.py (hunk 6 above)
Comment (Pringled) [reply model2vec/train/classifier.py:232]: I don't completely understand how to do this one. I just updated the code, what should I do to make this pass mypy without casting?

**Review by stephantul — APPROVED** [review]
